### PR TITLE
prov/rxm: Implement postponed queue for RMA operations

### PR DESCRIPTION
**Pull request by gladkovdmitry17**
- The 1st pathc applies some minor refactoring to RxM's RMA transport.
Some code is moved into common functions that will be re-used for implementing postponed queue for RMA
- The 2nd patch adds implementation of postponed queue for RMA operations

Signed-off-by: Dmitry Gladkov <dmitry.gladkov@intel.com>

### CODE REVIEW

File: prov/rxm/src/rxm_ep.c [hunk 1]
```diff
@@ -339,26 +339,37 @@ static int rxm_ep_txrx_res_open(struct rxm_ep *rxm_ep)
...
 	dlist_init(&rxm_ep->repost_ready_list);
 
+	ret = rxm_buf_pool_create(rxm_ep,
+				  rxm_ep->msg_info->tx_attr->size,
+				  rxm_ep->rxm_info->tx_attr->inject_size +
+				  sizeof(struct rxm_rma_buf),
+				  &rxm_ep->buf_pools[RXM_BUF_POOL_RMA],
+				  RXM_BUF_POOL_RMA);
+	if (ret)
+		goto err5;
```
Comment (shefty) [inline prov/rxm/src/rxm_ep.c]: This function needs to go on a diet.  Please consider moving all the pool allocations into a new function.

File: prov/rxm/src/rxm.h [hunk 2]
```diff
@@ -250,6 +251,18 @@ struct rxm_tx_buf {
 	struct rxm_pkt pkt;
 };
 
+struct rxm_rma_buf {
+	/* Must stay at top */
+	struct rxm_buf hdr;
+
+	/* Used for non-inject transfers */
+	struct fi_msg_rma msg;
+	struct rxm_iov rxm_iov[RXM_IOV_LIMIT];
```
Comment (a-ilango) [inline prov/rxm/src/rxm.h]: Are these fields going to be used in subsequent patches? They don't seem to be getting used in this patch and the reason to have a separate rxm_rma_buf instead rxm_tx_buf is not obvious.

File: prov/rxm/src/rxm_ep.c (hunk 1 above)
Comment (gladkovdmitry17) [reply prov/rxm/src/rxm_ep.c]: :) Yep, should be refactored

File: prov/rxm/src/rxm.h (hunk 2 above)
Comment (gladkovdmitry17) [reply prov/rxm/src/rxm.h]: Yes, these fields are utilized in the new prepared commit (2nd commit)

**Comment by gladkovdmitry17** [comment]
@shefty @a-ilango The implementation of postponed queue for RMA operations is ready.

File: prov/rxm/src/rxm_rma.c [hunk 3]
```diff
@@ -169,27 +365,36 @@ static ssize_t
...
+		fastlock_acquire(&rxm_ep->util_ep.cmap->lock);
+		ret = rxm_ep_format_rma_inject_res(rxm_ep, total_size,
+						   flags, FI_WRITE, msg,
+						   &rma_buf, &tx_entry);
+		if (OFI_LIKELY(!ret)) {
+			rxm_conn = container_of(handle, struct rxm_conn, handle);
+			dlist_insert_tail(&tx_entry->postponed_entry,
+					  &rxm_conn->postponed_tx_list);
+		}
+		fastlock_release(&rxm_ep->util_ep.cmap->lock);
```
Comment (a-ilango) [inline prov/rxm/src/rxm_rma.c]: Can we wrap the acquire lock + insert into postponed_tx_list in a function? Also I think the formatting function called before acquiring the lock.

File: prov/rxm/src/rxm_rma.c [hunk 4]
```diff
@@ -63,6 +63,186 @@ rxm_ep_rma_reg_iov(struct rxm_ep *rxm_ep, const struct iovec *msg_iov,
...
+		uint64_t flags = ((tx_entry->flags & FI_INJECT) ?
+				  ((tx_entry->flags & ~FI_INJECT) |
+				   FI_COMPLETION) : tx_entry->flags);
+		(void) fi_writemsg(rxm_conn->msg_ep,
+				   &tx_entry->rma_buf->msg,
+				   flags);
+	} else if (tx_entry->comp_flags & FI_READ) {
+		(void) fi_readmsg(rxm_conn->msg_ep,
+				  &tx_entry->rma_buf->msg,
+				  tx_entry->flags);
```
[the remark this answers is not in the document]
Comment (gladkovdmitry17) [reply prov/rxm/src/rxm_rma.c]: Sure, we should do this

File: prov/rxm/src/rxm_rma.c (hunk 3 above)
Comment (gladkovdmitry17) [reply prov/rxm/src/rxm_rma.c]: Yep, resources should be allocated outside the section where lock is already acquired
and yes, we can wrap.
Thanks!

File: prov/rxm/src/rxm_cq.c [hunk 5]
```diff
@@ -574,8 +574,12 @@ static ssize_t rxm_cq_handle_comp(struct rxm_ep *rxm_ep,
 			rxm_ep_msg_mr_closev(tx_entry->mr, tx_entry->count);
 		return rxm_finish_send_nobuf(tx_entry);
 	case RXM_TX:
-		assert(comp->flags & (FI_SEND | FI_WRITE));
+		assert(comp->flags & (FI_SEND));
 		return rxm_finish_send(tx_entry);
+	case RXM_TX_RMA:
+		assert(comp->flags & (FI_SEND));
```
[the remark this answers is not in the document]
Comment (a-ilango) [reply prov/rxm/src/rxm_cq.c]: Actually, for FI_READ the tx_entry state would be RXM_TX_NOBUF and not RXM_TX_RMA. That's what the old code did. I think the new code does the same as well.

[the remark this answers is not in the document]
Comment (gladkovdmitry17) [reply prov/rxm/src/rxm_cq.c]: Sorry, I forgot to provide some clarifications -- for FI_READ the `tx_entry::state` can be `RXM_TX_RMA` for only postponed case.  It uses iovec array (field of rxm_rma_buf) and stores user's iovec there.

[the remark this answers is not in the document]
Comment (shefty) [reply prov/rxm/src/rxm_cq.c]: I'm looking at the first commit by itself,  Based on that, this removes FI_WRITE from RXM_TX, so I would expect the new case to use FI_WRITE rather than FI_SEND, which agrees with Arun's initial comment.

File: prov/rxm/src/rxm_cq.c [hunk 6]
```diff
@@ -577,7 +579,7 @@ static ssize_t rxm_cq_handle_comp(struct rxm_ep *rxm_ep,
 		assert(comp->flags & (FI_SEND));
 		return rxm_finish_send(tx_entry);
 	case RXM_TX_RMA:
-		assert(comp->flags & (FI_SEND));
+		assert(comp->flags & (FI_WRITE | FI_READ));
```
Comment (shefty) [inline prov/rxm/src/rxm_cq.c]: Please make the necessary changes to the prior patch, which sets this.  It's not obvious why FI_READ is needed based on the original code.  Maybe that needs to be in a separate patch?

File: prov/rxm/src/rxm_rma.c [hunk 7]
```diff
@@ -37,7 +37,7 @@ typedef ssize_t rxm_rma_msg_fn(struct fid_ep *ep_fid,
 
 static inline ssize_t
 rxm_ep_rma_reg_iov(struct rxm_ep *rxm_ep, const struct iovec *msg_iov,
-		   void **desc, void **mr_storage,
+		   void ***desc, void **mr_storage,
```
Comment (shefty) [inline prov/rxm/src/rxm_rma.c]: I'm not sure mere mortals can process a pointer to a pointer to a pointer, especially when the pointer to the pointer to the pointer is set to point to the pointer to the pointer that's also passed in as a parameter.  (See what I mean? :)  We need _something_ a little simpler here.

File: prov/rxm/src/rxm_rma.c [hunk 8]
```diff
@@ -204,7 +256,7 @@ rxm_ep_format_rma_non_inject_res(struct rxm_ep *rxm_ep, size_t total_size,
 
 	ret = rxm_ep_rma_reg_iov(rxm_ep, (*rma_buf)->rxm_iov.iov,
 				 /* addr of desc from rma_buf will be assign to itself */
-				 (*rma_buf)->rxm_iov.desc, (*rma_buf)->rxm_iov.desc,
+				 (void ***)&(*rma_buf)->rxm_iov.desc, (*rma_buf)->rxm_iov.desc,
```
Comment (shefty) [inline prov/rxm/src/rxm_rma.c]: This is, odd.

File: prov/rxm/src/rxm_rma.c (hunk 3 above)
Comment (a-ilango) [reply prov/rxm/src/rxm_rma.c]: Is it still not possible to wrap the locking into rxm_ep_postponed_rma function? I think they are needed only around the `dlist_insert()` and once before to check the state and return if needed.

File: prov/rxm/src/rxm_rma.c [hunk 9]
```diff
@@ -219,6 +271,34 @@ rxm_ep_format_rma_non_inject_res(struct rxm_ep *rxm_ep, size_t total_size,
 	return ret;
 }
 
+static inline int
+rxm_ep_postponed_rma(struct rxm_ep *rxm_ep, struct rxm_conn *rxm_conn,
```
Comment (a-ilango) [inline prov/rxm/src/rxm_rma.c]: nit: this can be renamed to `rxm_ep_postpone_rma` to better reflect the intent.

File: prov/rxm/src/rxm_rma.c (hunk 3 above)
Comment (gladkovdmitry17) [reply prov/rxm/src/rxm_rma.c]: Sorry, it is not possible due to the following scenario:
1. main thread (MT) enters to rxm_ep_send_common()
2. (MT) tries to get connection handle (this is unsuccessful -- connection wasn't established yet)
3. (MT) during resources allocation to postpone operation -- connection thread (CT) acquires the lock has got successful connection establishment and then yields the lock.
4. (MT) acquires the lock and inserts the TX entry to the list of deferred operations (this will never be sent)

Comment (a-ilango) [reply prov/rxm/src/rxm_rma.c]: There has to be a check in the pospone_send function where it would check the connection state before inserting the op into the list. That should prevent any deferred op from not being sent. I would prefer this if there is a way to re-use the resources allocated in postpone_send for regular send as well (in case we find the connection has been established).

Nevertheless, I'm fine with the current code and the refactoring can be done later (if possible).

P.S. Github is hiding comments on outdated diff. It took me a while to find this.

File: prov/rxm/src/rxm_rma.c [hunk 10]
```diff
@@ -52,17 +52,253 @@ rxm_ep_rma_reg_iov(struct rxm_ep *rxm_ep, const struct iovec *msg_iov,
...
+			   struct rxm_tx_entry *tx_entry,
+			   const struct fi_msg_rma *orig_msg)
+{
+    uint8_t i;
+
+	rma_buf->rxm_iov.count = (uint8_t)orig_msg->iov_count;
+	for (i = 0; i < rma_buf->rxm_iov.count; i++) {
+		rma_buf->rxm_iov.iov[i].iov_base = orig_msg->msg_iov[i].iov_base;
+		rma_buf->rxm_iov.iov[i].iov_len = orig_msg->msg_iov[i].iov_len;
+	}
```
Comment (shefty) [inline prov/rxm/src/rxm_rma.c]: Can this be a memcpy?

File: prov/rxm/src/rxm_rma.c [hunk 11]
```diff
@@ -52,17 +52,253 @@ rxm_ep_rma_reg_iov(struct rxm_ep *rxm_ep, const struct iovec *msg_iov,
...
+				    rma_buf, *tx_entry);
+	if (OFI_UNLIKELY(!ret))
+		goto err2;
+
+	return FI_SUCCESS;
+err2:
+	rxm_tx_entry_release(&rxm_ep->send_queue, *tx_entry);
+err1:
+	FI_WARN(&rxm_prov, FI_LOG_CQ, "Unable to allocate resources!\n");
+	rxm_ep_progress_multi(&rxm_ep->util_ep);
```
Comment (shefty) [inline prov/rxm/src/rxm_rma.c]: This looks like a side effect for a function called 'format_rma_res'.

File: prov/rxm/src/rxm_rma.c [hunk 12]
```diff
@@ -71,32 +307,48 @@ rxm_ep_rma_common(struct rxm_ep *rxm_ep, const struct fi_msg_rma *msg, uint64_t
...
 	int ret;
 
 	ret = ofi_cmap_get_handle(rxm_ep->util_ep.cmap, msg->addr, &handle);
-	if (OFI_UNLIKELY(ret))
-		return ret;
+	if (OFI_UNLIKELY(ret)) {
+		if (OFI_UNLIKELY(ret != -FI_EAGAIN))
+			return ret;
+		fastlock_acquire(&rxm_ep->util_ep.cmap->lock);
+		if (handle->state != CMAP_CONNECTED) {
```
Comment (shefty) [inline prov/rxm/src/rxm_rma.c]: This flow looks odd.  ofi_cmap_get_handle() fails, but we still expect handle to be valid.

File: prov/rxm/src/rxm_rma.c [hunk 13]
```diff
@@ -169,85 +421,65 @@ static ssize_t
...
 	ssize_t ret;
 
 	ret = ofi_cmap_get_handle(rxm_ep->util_ep.cmap, msg->addr, &handle);
-	if (OFI_UNLIKELY(ret))
-		return ret;
+	if (OFI_UNLIKELY(ret)) {
+		if (OFI_UNLIKELY(ret != -FI_EAGAIN))
+			return ret;
+		fastlock_acquire(&rxm_ep->util_ep.cmap->lock);
+		if (handle->state != CMAP_CONNECTED) {
```
Comment (shefty) [inline prov/rxm/src/rxm_rma.c]: See above comment

File: prov/rxm/src/rxm_rma.c (hunk 10 above)
Comment (gladkovdmitry17) [reply prov/rxm/src/rxm_rma.c]: Ditto! Thanks

File: prov/rxm/src/rxm_rma.c (hunk 11 above)
Comment (gladkovdmitry17) [reply prov/rxm/src/rxm_rma.c]: Sorry, I'm not sure that I got it correct. What do you mean? Do you mean that `err1` label is excessive?

File: prov/rxm/src/rxm_rma.c (hunk 12 above)
Comment (gladkovdmitry17) [reply prov/rxm/src/rxm_rma.c]: Yes, I agree. But we should handle case here that connection is successfully established in CM thread when trying to hold `cmap::lock` lock from master thread.
The another way is to remove acquiring of `cmap::lock` lock and each calling of `ofi_cmap_get_handle` has to hold `cmap::lock`... Not sure that this is right way

File: prov/rxm/src/rxm_rma.c (hunk 11 above)
Comment (shefty) [reply prov/rxm/src/rxm_rma.c]: The name of the function suggests that all this call does is to format some sort of request.  Calling 'progress' in the case of a failure comes across as an unexpected operation.  I didn't see this in the other 'format' calls.

File: prov/rxm/src/rxm_rma.c (hunk 12 above)
Comment (shefty) [reply prov/rxm/src/rxm_rma.c]: I think the best option for this code flow (and any that are similar) is to separate ofi_cmap_get_handle() into two calls -- cmap_acquire_handle() and handle_connect() -- but with better names.
```
/* top half of ofi_cmap_get_handle */
ofi_cmap_acquire_handle(...)
{
    handle = util_cmap_get_handle(...)
    if (!handle)
        util_cmap_alloc_handle(..., &handle);
    return handle
}

/* bottom half of ofi_cmap_get_handle */
ofi_handle_connect(...)
{
    if (handle->state == CMAP_CONNECT)
        return 0;
    switch (handle->state) {
        ...
    }
    return -FI_EAGAIN;
}
```
This code can then be restructured to avoid the double lock acquire-release flow.  E.g.
```
lock(...)
handle = ofi_cmap_acquire_handle(...);
if (handle && handle->state != CMAP_CONNECTED) {
    ...
}
```
This needs more work, but that's the general idea.

Comment (shefty) [reply prov/rxm/src/rxm_rma.c]: I should have added that ofi_cmap_get_handle() would be modified to call these 2 new functions, in case that wasn't obvious.

Comment (a-ilango) [reply prov/rxm/src/rxm_rma.c]: We could also add a callback to ofi_cmap_get_handle which would be called if the connection is not established yet. The callback can then add the send to postponed list. That would keep the lock-release code confined to cmap.

File: prov/rxm/src/rxm_rma.c (hunk 11 above)
Comment (gladkovdmitry17) [reply prov/rxm/src/rxm_rma.c]: Oh, I got it. Thanks. Removed this call

File: prov/rxm/src/rxm_rma.c (hunk 12 above)
Comment (gladkovdmitry17) [reply prov/rxm/src/rxm_rma.c]: Thanks! This makes sense.
I've created separate PR #3879 that addresses this comment for send/inject calls

@a-ilango it would be great to have this, but unfortunately we should build arg to pass them into `ofi_cmap_get_handle()` before calling this. This adds some overhead

**Comment by shefty** [comment]
#3879 was merged - it would be great to update this PR based on those changes and have this merged.  We're targeting the 1.6 release in 2 weeks.  I can create an rc3 in a few more days if we have other changes to fit into 1.6

**Comment by gladkovdmitry17** [comment]
> it would be great to update this PR based on those changes and have this merged.

Let me 20 minutes :)

**Comment by gladkovdmitry17** [comment]
> it would be great to update this PR based on those changes and have this merged.
> 
> Let me 20 minutes :)

@shefty @a-ilango Done
